### PR TITLE
Annotated the text package

### PR DESCRIPTION
**Pull request by Kakarot-SSJ4**

### CODE REVIEW

**Comment by Kakarot-SSJ4** [comment]
@kelloggm @panacekcz

**Review by kelloggm — APPROVED** [review]
I'm conditionally approving this pull request. Please merge the differences between this PR and #2 so that the only changes in the this PR are the changes to the text package. Once you've done that, you should tag @panacekcz, and he can review this PR.

**Comment by Kakarot-SSJ4** [comment]
@panacekcz

File: src/main/java/org/apache/commons/lang3/text/StrBuilder.java [hunk 1]
```diff
@@ -738,6 +782,7 @@ public StrBuilder append(final StringBuffer str, final int startIndex, final int
      * @return this, to enable chaining
      * @since 3.2
      */
+@SuppressWarnings({"index:compound.assignment.type.incompatible", "index:argument.type.incompatible"}) // #1, #2: ensureCapacity(len + length) => len + length is @NonNegative @LTEqLengthOf("this.buffer"), and len is @IndexOrHigh("this.buffer") as well
```
Comment (panacekcz) [inline src/main/java/org/apache/commons/lang3/text/StrBuilder.java]: Indentation.

File: src/main/java/org/apache/commons/lang3/text/StrBuilder.java [hunk 2]
```diff
@@ -837,6 +891,10 @@ public StrBuilder append(final StrBuilder str, final int startIndex, final int l
      * @param chars  the char array to append
      * @return this, to enable chaining
      */
+     @SuppressWarnings({"index:compound.assignment.type.incompatible", "index:argument.type.incompatible"}) /*
```
Comment (panacekcz) [inline src/main/java/org/apache/commons/lang3/text/StrBuilder.java]: Indentation.

File: src/main/java/org/apache/commons/lang3/text/StrSubstitutor.java [hunk 3]
```diff
@@ -418,7 +424,7 @@ public String replace(final String source) {
      * @param length  the length within the array to be processed, must be valid
      * @return the result of the replace operation
      */
-    public String replace(final String source, final int offset, final int length) {
+    public String replace(final String source, final @IndexFor("#1") int offset, final @NonNegative @LTLengthOf(value = {"#1"}, offset = {"#2 - 1"}) int length) {
```
Comment (panacekcz) [inline src/main/java/org/apache/commons/lang3/text/StrSubstitutor.java]: Can offset be `IndexOrHigh`? Usually when a method has an offset and length parameter, having `offset == array.length` and `length == 0` is valid.

File: src/main/java/org/apache/commons/lang3/text/StrSubstitutor.java [hunk 4]
```diff
@@ -460,7 +466,7 @@ public String replace(final char[] source) {
      * @param length  the length within the array to be processed, must be valid
      * @return the result of the replace operation
      */
-    public String replace(final char[] source, final int offset, final int length) {
+    public String replace(final char[] source, final int offset, final @NonNegative int length) {
```
Comment (panacekcz) [inline src/main/java/org/apache/commons/lang3/text/StrSubstitutor.java]: Should this method be annotated similarly to `replace(String, int, int)`? (And the same for other similar methods in this class.)

File: src/main/java/org/apache/commons/lang3/text/StrSubstitutor.java [hunk 5]
```diff
@@ -635,7 +642,7 @@ public boolean replaceIn(final StringBuffer source) {
      * @param length  the length within the buffer to be processed, must be valid
      * @return true if altered
      */
-    public boolean replaceIn(final StringBuffer source, final int offset, final int length) {
+    public boolean replaceIn(final StringBuffer source, final @IndexFor("#1") int offset, final @NonNegative @LTLengthOf(value = {"#1"}, offset = {"#2 - 1"}) int length) {
```
Comment (panacekcz) [inline src/main/java/org/apache/commons/lang3/text/StrSubstitutor.java]: Again, can offset be `IndexOrHigh`?

File: src/main/java/org/apache/commons/lang3/text/StrTokenizer.java [hunk 6]
```diff
@@ -407,6 +412,7 @@ public int size() {
      *
      * @return the next sequential token, or null when no more tokens are found
      */
+    @SuppressWarnings({"index:array.access.unsafe.high", "index:unary.increment.type.incompatible"}) // hasNext() => tokenPos < tokens.length
```
Comment (panacekcz) [inline src/main/java/org/apache/commons/lang3/text/StrTokenizer.java]: Can this be expressed by a conditional postcondition (`@EnsuresLTLengthOfIf`) on `hasNext`?

File: src/main/java/org/apache/commons/lang3/text/StrBuilder.java [hunk 7]
```diff
@@ -1666,6 +1755,14 @@ public StrBuilder insert(final int index, final char chars[]) {
...
      */
+    @SuppressWarnings({"index:argument.type.incompatible", "index:compound.assignment.type.incompatible"}) /*
+    #1: validateIndex(index) implies index to be @IndexOrHigh("buffer"), but ensureCapacity(size + length)
+        for length > 0 increases the size of buffer hence making index to be @IndexFor("buffer")
+        ensureCapacity(size + length) for length > 0 => size - index to be @NonNegative,
+        also size - index <= buffer.length - index - length because of ensureCapacity(size + length)
+    #2: offset >= 0, and offset + length <= chars.length, but length > 0, hence offset < chars.length
+    #3: ensureCapacity(size + length) => size + length is @LTEqLengthOf("this.buffer")
+    */
     public StrBuilder insert(final int index, final char chars[], final int offset, final int length) {
```
Comment (panacekcz) [inline src/main/java/org/apache/commons/lang3/text/StrBuilder.java]: Looks like `offset` should be `@IndexOrHigh("#2")`.

File: src/main/java/org/apache/commons/lang3/text/StrTokenizer.java [hunk 8]
```diff
@@ -828,9 +846,10 @@ private int readWithQuotes(final char[] srcChars, final int start, final int len
      * @param quoteLen  the length of the matched quote, 0 if no quoting
      * @return true if a quote is matched
      */
+    @SuppressWarnings({"index:array.access.unsafe.low", "index:array.access.unsafe.high"}) // #1: If pos + i >= len, srcChars[pos + i] and srcChars[quoteStart + i] does not happen. Also, pos >= quoteStart as is checked when it is called. This function is called only in readWithQuotes() which is called only by readNextTokens(), which declares pos and quotestart to either be start + quoteLen and start respectively or start and 0 respectively, where start and quoteLen both are @NonNegative.
```
Comment (panacekcz) [inline src/main/java/org/apache/commons/lang3/text/StrTokenizer.java]: Why not annotate the parameters with `@NonNegative`?

**Review by panacekcz — COMMENTED** [review]
The comments are very good, but there are places where annotations seem to be missing, or there are very similar methods inconsistently annotated.

**Comment by Kakarot-SSJ4** [comment]
@panacekcz made the suggested changes

**Pull request merged**